compiler/rustc_trait_selection/src/traits/select/mod.rs

```diff
@@ -1306,12 +1306,6 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
         dep_node: DepNodeIndex,
         result: EvaluationResult,
     ) {
-        // Avoid caching results that depend on more than just the trait-ref
-        // - the stack can create recursion.
-        if result.is_stack_dependent() {
-            return;
```

Comment:
How is it sound to remove this? Aren't we caching things that should not be cached?